core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,398 @@
         total_answers_count: 0,
         useful_feedback_count: 0,
         total_hit_count: 0,
         first_hit_count: 0,
         num_times_solution_viewed: 0,
         num_completions: 0
       };
-    };
+    }
 
-    var getFullStatsUrl = function(urlIdentifier) {
+    private getFullStatsUrl(urlIdentifier: string): string {
       try {
-        return UrlInterpolationService.interpolateUrl(
-          STATS_REPORTING_URLS[urlIdentifier], {
-            exploration_id: explorationId
+        return this.urlInterpolationService.interpolateUrl(
+          ExplorationPlayerConstants.STATS_REPORTING_URLS[urlIdentifier], {
+            exploration_id: this.explorationId
           });
       } catch (e) {
         var additionalInfo = ('\nUndefined exploration id error debug logs:' +
           '\nThe event being recorded: ' + urlIdentifier +
-          '\nExploration ID: ' + ContextService.getExplorationId()
+          '\nExploration ID: ' + this.contextService.getExplorationId()
         );
-        if (currentStateName) {
-          additionalInfo += ('\nCurrent State name: ' + currentStateName);
+        if (this.currentStateName) {
+          additionalInfo += ('\nCurrent State name: ' + this.currentStateName);
         }
-        if (nextExpId) {
-          additionalInfo += ('\nRefresher exp id: ' + nextExpId);
+        if (this.nextExpId) {
+          additionalInfo += ('\nRefresher exp id: ' + this.nextExpId);
         }
-        if (previousStateName && nextStateName) {
-          additionalInfo += ('\nOld State name: ' + previousStateName +
-            '\nNew State name: ' + nextStateName);
+        if (this.previousStateName && this.nextStateName) {
+          additionalInfo += ('\nOld State name: ' + this.previousStateName +
+            '\nNew State name: ' + this.nextStateName);
         }
         e.message += additionalInfo;
         throw e;
       }
-    };
+    }
 
-    var startStatsTimer = function() {
-      if (!_editorPreviewMode && !_questionPlayerMode ) {
-        $interval(function() {
-          postStatsToBackend();
-        }, 300000);
+    private startStatsTimer(): void {
+      if (!this._editorPreviewMode && !this._questionPlayerMode ) {
+        const secondsCounter = interval(300000);
+        secondsCounter.subscribe(n => this.postStatsToBackend());
       }
-    };
+    }
 
     // This method is called whenever a learner tries to leave an exploration,
     // when a learner starts an exploration, when a learner completes an
     // exploration and also every five minutes.
-    var postStatsToBackend = function() {
-      if (explorationIsComplete) {
+    private postStatsToBackend(): void {
+      if (this.explorationIsComplete) {
         return;
       }
-      $http.post(getFullStatsUrl('STATS_EVENTS'), {
-        aggregated_stats: aggregatedStats,
-        exp_version: explorationVersion
+      this.http.post(this.getFullStatsUrl('STATS_EVENTS'), {
```

Comment:
Ditto: add TODO for moving this into a `backend-api.service` file.

Reply:
Done.